File: longevity/measures.py
"""Виды измерений биодневника с единицами и форматированием."""

from dataclasses import dataclass


@dataclass(frozen=True)
class Kind:
    id: str
    title: str
    unit: str


KINDS = (
    Kind("weight", "Вес", "кг"),
    Kind("waist", "Окружность талии", "см"),
    Kind("bp_sys", "Давление систолическое", "мм рт.ст."),
    Kind("bp_dia", "Давление диастолическое", "мм рт.ст."),
    Kind("pulse", "Пульс", "уд/мин"),
    Kind("walk6m", "Тест 6-минутной ходьбы", "м"),
    Kind("grip", "Сила хвата", "кг"),
)


def kind(kind_id: str) -> Kind | None:
    for k in KINDS:
        if k.id == kind_id:
            return k
    return None


def format_value(kind_id: str, value: float) -> str:
    """Текст значения с единицей: «72.5 кг», «80 кг»."""
    k = kind(kind_id)
    unit = k.unit if k else ""
    if value == int(value):
        number = str(int(value))
    else:
        number = f"{value:.1f}"
    return f"{number} {unit}".strip() if unit else number


def delta(kind_id: str, current: float, previous: float) -> str:
    """Изменение относительно предыдущего: «+0.4 кг», «−1.2 кг» или «0.0 кг»."""
    diff = current - previous
    if diff > 0.049:
        sign = "+"
    elif diff < -0.049:
        sign = "−"
    else:
        sign = ""
    k = kind(kind_id)
    unit = k.unit if k else ""
    return f"{sign}{abs(diff):.1f} {unit}".strip()

File: android/app/src/main/java/com/revaks/longevity/core/Measures.kt
package com.revaks.longevity.core

/** Виды измерений биодневника с единицами. */
object Measures {

    data class Kind(val id: String, val title: String, val unit: String)

    val KINDS: List<Kind> = listOf(
        Kind("weight", "Вес", "кг"),
        Kind("waist", "Окружность талии", "см"),
        Kind("bp_sys", "Давление систолическое", "мм рт.ст."),
        Kind("bp_dia", "Давление диастолическое", "мм рт.ст."),
        Kind("pulse", "Пульс", "уд/мин"),
        Kind("walk6m", "Тест 6-минутной ходьбы", "м"),
        Kind("grip", "Сила хвата", "кг"),
    )

    fun kind(id: String): Kind? = KINDS.firstOrNull { it.id == id }

    /** Текст значения с единицей: «72.5 кг». */
    fun format(kindId: String, value: Double): String {
        val unit = kind(kindId)?.unit ?: ""
        val number = if (value == value.toLong().toDouble()) {
            value.toLong().toString()
        } else {
            String.format(java.util.Locale.US, "%.1f", value)
        }
        return if (unit.isEmpty()) number else "$number $unit"
    }

    /** Изменение относительно предыдущего значения: «+0.4 кг», «−1.2 кг» или «0.0 кг». */
    fun delta(kindId: String, current: Double, previous: Double): String {
        val diff = current - previous
        val sign = when {
            diff > 0.049 -> "+"
            diff < -0.049 -> "−"
            else -> ""
        }
        val unit = kind(kindId)?.unit ?: ""
        val number = String.format(java.util.Locale.US, "%.1f", kotlin.math.abs(diff))
        return "$sign$number $unit".trim()
    }
}
